python-checks/src/main/java/org/sonar/python/checks/hotspots/GraphQLIntrospectionCheck.java

```diff
@@ -0,0 +1,123 @@
+    "graphene.validation.DisableIntrospection",
+    "graphql.validation.NoSchemaIntrospectionCustomRule");
+
+  private static final String MESSAGE = "Disable introspection on this GraphQL server endpoint.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, GraphQLIntrospectionCheck::checkGraphQLIntrospection);
+  }
+
+  private static void checkGraphQLIntrospection(SubscriptionContext ctx) {
+    CallExpression callExpression = (CallExpression) ctx.syntaxNode();
+    Optional.ofNullable(callExpression.calleeSymbol())
+      .map(Symbol::fullyQualifiedName)
+      .filter(fqn -> GRAPHQL_VIEWS_FQNS.contains(fqn) &&
+        (hasSensitiveMiddlewares(callExpression.arguments()) || hasSensitiveValidationRules(callExpression.arguments())))
+      .ifPresent(fqn -> ctx.addIssue(callExpression, MESSAGE));
+  }
+
+  private static boolean hasSensitiveMiddlewares(List<Argument> arguments) {
+    return areArgumentsUnsafe("middleware", expressionsNameContainIntrospection, arguments);
+  }
+
+  private static boolean hasSensitiveValidationRules(List<Argument> arguments) {
+    return areArgumentsUnsafe("validation_rules",
+      expressionsNameContainIntrospection.or(expressionsContainsSensitiveRuleFQN),
+      arguments);
+  }
+
+  private static boolean areArgumentsUnsafe(String argumentName, Predicate<List<Expression>> check, List<Argument> arguments) {
+    var middlewareArgument = TreeUtils.argumentByKeyword(argumentName, arguments);
+    if (middlewareArgument == null) {
+      return true;
+    }
+    return Optional.of(middlewareArgument)
+      .map(RegularArgument::expression)
+      .flatMap(GraphQLIntrospectionCheck::expressionsFromListOrTuple)
+      .map(expressions -> expressions.isEmpty() || check.test(expressions))
+      .orElse(false);
+  }
+
+  private static Optional<List<Expression>> expressionsFromListOrTuple(Expression expression) {
+    return TreeUtils.toOptionalInstanceOfMapper(ListLiteral.class).apply(expression)
+      .map(ListLiteral::elements)
+      .map(ExpressionList::expressions)
+      .or(() -> TreeUtils.toOptionalInstanceOfMapper(Tuple.class)
+        .apply(expression)
+        .map(Tuple::elements));
+  }
+
+  private static Predicate<List<Expression>> expressionsNameContainIntrospection = expressions -> expressions.stream()
+    .map(GraphQLIntrospectionCheck::nameFromIndentifierOrCallExpression)
+    .filter(Optional::isPresent)
+    .map(Optional::get)
+    .map(String::toUpperCase)
+    .anyMatch(name -> name.contains("INTROSPECTION"));
+
+  private static Predicate<List<Expression>> expressionsContainsSensitiveRuleFQN = expressions -> expressions.stream()
+    .map(TreeUtils::getSymbolFromTree)
+    .filter(Optional::isPresent)
+    .map(Optional::get)
+    .map(Symbol::fullyQualifiedName)
+    .filter(Objects::nonNull)
+    .anyMatch(SENSITIVE_VALIDATION_RULE_FQNS::contains);
```

Comment:
I like functional API in Java, and I'm trying to use it as much as possible but even for me it took some time to figure out what's going on here. It is up to you but maybe it will be easier to read the code if you will have just two methods to check the expressions, like:
```java
private static boolean isExpressionsNameContainIntrospection(List<Expression> expressions) {
    return expressions.stream()
      .map(GraphQLIntrospectionCheck::nameFromIndentifierOrCallExpression)
      .filter(Optional::isPresent)
      .map(Optional::get)
      .map(String::toUpperCase)
      .anyMatch(name -> name.contains("INTROSPECTION"));
  }
  
  private static boolean isExpressionsContainsSensitiveRuleFQN(List<Expression> expressions) {
    return expressions.stream()
      .map(TreeUtils::getSymbolFromTree)
      .filter(Optional::isPresent)
      .map(Optional::get)
      .map(Symbol::fullyQualifiedName)
      .filter(Objects::nonNull)
      .anyMatch(SENSITIVE_VALIDATION_RULE_FQNS::contains);
  }
  
  private static boolean isExpressionsNameContainIntrospectionOrContainsSensitiveRuleFQN(List<Expression> expressions) {
    return isExpressionsNameContainIntrospection(expressions) || isExpressionsContainsSensitiveRuleFQN(expressions);
  }
```

Reply:
I see so remove the `or` on line 75. And I guess not use a predicate in the `areArgumentsStafe` method

Reply:
With the change I had a bit more duplicate code as I had to treat empty list and empty `Optional` differently but I think the intent is quite clear now.